import os
import yaml
from collections import defaultdict

### FUNCTIONS ###

def map_input(wildcards):
    inputs=[]
    for RUN,_run in list(FILES[wildcards.sample].items()):
        run,lane,index=_run['PU'].split('-',2)
        inputs.append(f'bam_input/work/{wildcards.sample}/{wildcards.reference}/{run}/{lane}/{index}/disambiguated.bam')
    assert len(inputs)>0
    return sorted(inputs)

def get_fastqs(wildcards):
    return {'R1':'FASTQ/'+FILES[wildcards.sample][f'{wildcards.run}-{wildcards.lane}']['files'][0],'R2':'FASTQ/'+FILES[wildcards.sample][f'{wildcards.run}-{wildcards.lane}']['files'][1]}
    #fastq.yaml does not currently include file paths.

### ### PYTHON ### ###

with open(config['project']['fastq_config']) as file:
    FILES=yaml.load(file,Loader=yaml.BaseLoader)
    SAMPLES=sorted(list(FILES.keys()))
    for sample in SAMPLES:
        os.makedirs(f'logs/cluster/{sample}',exist_ok=True)

### ### ### RULES ### ### ###
#Consider making bams.table the desired result.

rule all:
    input:
        expand("bam_input/final/{sample}/{reference}/{sample}.bam",sample=SAMPLES,reference=config['reference']['key'])

#check if bbmap is installed?

rule bbsplit:
    input:
        unpack(get_fastq)
    output:
        out_x=WM4023_S31285117_FGC2238_1_GAATTCGT-ATAGAGGC.GRCh37.fastq.gz,
        out_y=WM4023_S31285117_FGC2238_1_GAATTCGT-ATAGAGGC.mm10.fastq.gz,
        refstats="metrics/{sample}/"
    params:
        ref_p="/home/bwubb/resources/Genomes/PDX/"
    threads:
        8
    shell:
        """
        bbsplit.sh path={params.ref_p} in1={input.R1} in2={input.R2} out_x= out_y= refstats=
        """

rule human_aln_pe:#speciesA
    input:
        unpack(get_fastqs)
    output:
        temp("bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/mapped.bam")
    params:
        LB=config['resources']['library_key'],
        fasta=config['reference']['fasta']
    threads:
        8
    shell:
        "bwa mem -M -t {threads} {params.fasta} {input.R1} {input.R2} | samtools addreplacerg -r 'ID:{wildcards.run}.{wildcards.lane}' -r 'PU:{wildcards.run}.{wildcards.lane}.{wildcards.index}' -r 'PL:illumina' -r 'LB:{params.LB}' -r 'SM:{wildcards.sample}' -@ {threads} - | samtools sort -n -@ {threads} -o {output}"

rule mouse_aln_pe:#species B
    input:
        unpack(get_fastqs)
    output:
        temp("bam_input/work/{sample}/mm10/{run}/{lane}/{index}/mapped.bam")
    params:
        LB=config['resources']['library_key'],
        fasta=config['disambiguate']['fasta']#ref_key could get the fastq
    threads:
        8
    shell:
        #fastq config has no path.
        "bwa mem -M -t {threads} {params.fasta} {input.R1} {input.R2} | samtools addreplacerg -r 'ID:{wildcards.run}.{wildcards.lane}' -r 'PU:{wildcards.run}.{wildcards.lane}.{wildcards.index}' -r 'PL:illumina' -r 'LB:{params.LB}' -r 'SM:{wildcards.sample}' -@ {threads} - | samtools sort -n -@ {threads} -o {output}"

#input has been changed so it is name sorted
rule disambiguate:
    input:
        A="bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/mapped.bam",
        B="bam_input/work/{sample}/mm10/{run}/{lane}/{index}/mapped.bam"
    output:
        A=temp("bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/mapped.disambiguatedSpeciesA.bam"),
        B=temp("bam_input/work/{sample}/mm10/{run}/{lane}/{index}/mapped.disambiguatedSpeciesB.bam")
    params:
        work_dir="bam_input/work/{sample}/disambres",
        A=temp("bam_input/work/{sample}/disambres/mapped.disambiguatedSpeciesA.bam"),
        B=temp("bam_input/work/{sample}/disambres/mapped.disambiguatedSpeciesB.bam")
    shell:
        """
        python disambiguate.py -i {params.work_dir} -o {params.work_dir} -a bwa --no-sort {input.A} {input.B}
        rsync {params.A} {output.A}
        #samtools index {output.A}
        rsync {params.B} {output.B}
        #samtools index {output.B}
        """
'''
#BBSplit
#Written by Brian Bushnell, from Dec. 2010 - present
#Last modified June 11, 2018

#Description:  Maps reads to multiple references simultaneously.
#Outputs reads to a file for the reference they best match, with multiple options for dealing with ambiguous mappings.

#To index:     bbsplit.sh build=<1> ref_x=<reference fasta> ref_y=<another reference fasta>
#To map:       bbsplit.sh build=<1> in=<reads> out_x=<output file> out_y=<another output file>

#To be concise, and do everything in one command:
#bbsplit.sh ref=x.fa,y.fa in=reads.fq basename=o%.fq

#that is equivalent to
#bbsplit.sh build=1 in=reads.fq ref_x=x.fa ref_y=y.fa out_x=ox.fq out_y=oy.fq

#By default paired reads will yield interleaved output, but you can use the # symbol to produce twin output files.
#For example, basename=o%_#.fq will produce ox_1.fq, ox_2.fq, oy_1.fq, and oy_2.fq.

#Indexing Parameters (required when building the index):
#ref=<file,file>     A list of references, or directories containing fasta files.
#ref_<name>=<ref.fa> Alternate, longer way to specify references. e.g., ref_ecoli=ecoli.fa
#                    These can also be comma-delimited lists of files; e.g., ref_a=a1.fa,a2.fa,a3.fa
#build=<1>           If multiple references are indexed in the same directory, each needs a unique build ID.
#path=<.>            Specify the location to write the index, if you dont want it in the current working directory.

#Input Parameters:
#build=<1>           Designate index to use.  Corresponds to the number specified when building the index.
#in=<reads.fq>       Primary reads input; required parameter.
#in2=<reads2.fq>     For paired reads in two files.
'''

#Verify I need this after I fixed no-sort issue
#bams SHOULD be name sorted and they were not
rule reformat_and_repair:
    input:
        "bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/mapped.disambiguatedSpeciesA.bam"
    output:
        R1=temp("bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/disambiguated_R1.fastq"),
        R2=temp("bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/disambiguated_R2.fastq")
    params:
        qsort=temp("bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/qSortA.bam")
    shell:
        """
        #samtools sort -n -o {params.qsort} {input}
        reformat.sh -Xmx5g in={input} out=stdout.fq |
        reformat.sh -Xmx5g int=t addcolon=t uniquenames=t in=stdin.fq out=stdout.fq |
        repair.sh -Xmx5g in=stdin.fq out1={output.R1} out2={output.R2}
        """
#add bwa if above benefits form multithread

#Note this one is coordinate sorted
rule aln_pe:
    input:
        R1="bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/disambiguated_R1.fastq",
        R2="bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/disambiguated_R2.fastq"
    output:
        temp("bam_input/work/{sample}/GRCh37/{run}/{lane}/{index}/disambiguated.bam")
    params:
        LB=config['resources']['library_key'],
        fasta=config['reference']['fasta']
    threads:
        8
    shell:
        "bwa mem -M -t {threads} {params.fasta} {input.R1} {input.R2} | samtools addreplacerg -r 'ID:{wildcards.run}.{wildcards.lane}' -r 'PU:{wildcards.run}.{wildcards.lane}.{wildcards.index}' -r 'PL:illumina' -r 'LB:{params.LB}' -r 'SM:{wildcards.sample}' -@ {threads} - | samtools sort -@ {threads} -o {output}"
    #picard --metrics_accumulation_level readgroup will look at PU and will only fall back to ID if it is null'
    #Consider putting markdup in this as well. It outputs stats and multi threads. Per lane stats would be nice and HsMetrics will also do that.
    #markdup would first need samtools fixmate -m

rule input_ready:
    input:
        map_input
    output:
        temp("bam_input/work/{sample}/{reference}/sort.bam")
    params:
        input=temp("bam_input/work/{sample}/{reference}/input.bam")
    run:
        if len(input)==1:
            shell("rsync {input} {params.input}")
            shell("samtools sort {params.input} -o {output}")
        else:
            shell("samtools merge -f {params.input} {input}")
            shell("samtools sort {params.input} -o {output}")

rule MarkDuplicates:
    input:
        "bam_input/work/{sample}/{reference}/sort.bam"
    output:
        bam=temp("bam_input/work/{sample}/{reference}/mDup.bam"),
        metrics="bam_input/final/{sample}/metrics/{reference}/mark_duplicates.table"
    params:
        memory="10240m"
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/picard/2.20.7/picard.jar MarkDuplicates I={input} O={output.bam} M={output.metrics} CREATE_INDEX=true VALIDATION_STRINGENCY=LENIENT"

rule RealignerTargetCreator:
    input:
        "bam_input/work/{sample}/{reference}/mDup.bam"
    output:
        "bam_input/work/{sample}/{reference}/IndelRealigner.intervals"
    params:
        memory="10240m",
        reference=config['reference']['fasta'],
        known=["$HOME/resources/gatk/1000G_phase1.indels.b37.vcf","$HOME/resources/gatk/Mills_and_1000G_gold_standard.indels.b37.vcf"]
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar -R {params.reference} -T RealignerTargetCreator -I {input} -o {output} -known {params.known[0]} -known {params.known[1]}"

rule IndelRealigner:
    input:
        bam="bam_input/work/{sample}/{reference}/mDup.bam",
        targets="bam_input/work/{sample}/{reference}/IndelRealigner.intervals"
    output:
        temp("bam_input/work/{sample}/{reference}/realign.bam")
    params:
        memory="10240m",
        reference=config['reference']['fasta'],
        known=["$HOME/resources/gatk/1000G_phase1.indels.b37.vcf","$HOME/resources/gatk/Mills_and_1000G_gold_standard.indels.b37.vcf"]
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar -R {params.reference} -T IndelRealigner -I {input.bam} -o {output} -targetIntervals {input.targets} -known {params.known[0]} -known {params.known[1]}"

rule FirstPass_BaseRecalibrator:#update resources
    input:
        "bam_input/work/{sample}/{reference}/realign.bam"
    output:
        "bam_input/final/{sample}/metrics/{reference}/recal_data.table"
    params:
        memory="10240m",
        reference=config['reference']['fasta'],
        knownSites=["$HOME/resources/gatk/dbsnp_138.b37.vcf","$HOME/resources/gatk/1000G_phase1.indels.b37.vcf","$HOME/resources/gatk/Mills_and_1000G_gold_standard.indels.b37.vcf"]
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar -R {params.reference} -T BaseRecalibrator -I {input} -o {output} -knownSites {params.knownSites[0]} -knownSites {params.knownSites[1]} -knownSites {params.knownSites[2]}"

rule SecondPass_BaseRecalibrator:
    input:
        "bam_input/work/{sample}/{reference}/realign.bam",
        "bam_input/final/{sample}/metrics/{reference}/recal_data.table"
    output:
        "bam_input/final/{sample}/metrics/{reference}/post_recal_data.table"
    params:
        memory="10240m",
        reference=config['reference']['fasta'],
        knownSites=["$HOME/resources/gatk/dbsnp_138.b37.vcf","$HOME/resources/gatk/1000G_phase1.indels.b37.vcf","$HOME/resources/gatk/Mills_and_1000G_gold_standard.indels.b37.vcf"]
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar -R {params.reference} -T BaseRecalibrator -I {input[0]} -BQSR {input[1]} -o {output} -knownSites {params.knownSites[0]} -knownSites {params.knownSites[1]} -knownSites {params.knownSites[2]}"

rule AnalyzeCovariates:
    input:
        before="bam_input/final/{sample}/metrics/{reference}/recal_data.table",
        after="bam_input/final/{sample}/metrics/{reference}/post_recal_data.table"
    output:
        csv="bam_input/final/{sample}/metrics/BQSR.csv",
        pdf="bam_input/final/{sample}/metrics/BQSR.pdf"
    params:
        memory="10240m",
        reference=config['reference']['fasta'],
        knownSites=["$HOME/resources/gatk/dbsnp_135.b37.vcf","$HOME/resources/gatk/1000G_phase1.indels.b37.vcf","$HOME/resources/gatk/Mills_and_1000G_gold_standard.indels.b37.vcf"]
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar -R {params.reference} -T AnalyzeCovariates -before {input.before} -after {input.after} -csv {output.csv} -plots {output.pdf}"

rule PrintReads:
    input:
        bam="bam_input/work/{sample}/{reference}/realign.bam",
        bqsr="bam_input/final/{sample}/metrics/{reference}/recal_data.table"
    output:
        temp("bam_input/work/{sample}/{reference}/recal.bam")
    params:
        memory="1020m",
        reference=config['reference']['fasta']
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/GenomeAnalysisTK-3.7/GenomeAnalysisTK.jar -R {params.reference} -T PrintReads -I {input.bam} -BQSR {input.bqsr} -o {output}"

rule ValidateSamFile:
    #Errors and Warnings trigger non-zero exit
    input:
        "bam_input/work/{sample}/{reference}/recal.bam"
    output:
        "bam_input/work/{sample}/{reference}/validation_data.table"
    params:
        memory="10240m"
    shell:
        "java -Xmx{params.memory} -jar $HOME/software/picard/2.20.7/picard.jar ValidateSamFile I={input} O={output} MODE=SUMMARY"

rule validation_pass:
    input:
        "bam_input/work/{sample}/{reference}/validation_data.table"
    output:
        "bam_input/final/{sample}/{reference}/validation_data.table"
    shell:
        """
        set +H
        if egrep -q 'No errors found' {input[0]}; then
            cp {input[0]} {output[0]}
        else
            egrep '^ERROR' {input[0]}
            exit 1
        fi
        """

rule ready_bam:
    input:
        bam="bam_input/work/{sample}/{reference}/recal.bam",
        table="bam_input/final/{sample}/{reference}/validation_data.table"
    output:
        "bam_input/final/{sample}/{reference}/{sample}.bam"
    shell:
        """
        rsync -v {input.bam} {output}
        samtools index {output}
        """


